packages/js-dapi-client/lib/BlockHeadersProvider/BlockHeadersProvider.js

```diff
@@ -55,67 +74,197 @@ class BlockHeadersProvider extends EventEmitter {
+  /**
+   * @private
+   */
+  async init() {
+    if (!this.wasmX11Ready) {
+      const x11Hash = await X11();
+      configureDashcore({
+        x11hash: x11Hash,
+      });
+      this.wasmX11Ready = true;
     }
 
-    if (this.started) {
-      throw new Error('BlockHeaderProvider has already been started');
-    }
+    if (!this.blockHeadersReader) {
+      const createContinuousSyncStream = (fromBlockHeight) => DAPIStream
+        .create(
+          this.coreMethods.subscribeToBlockHeadersWithChainLocks,
+        )({
+          fromBlockHeight,
+        });
 
-    const { chain: { blocksCount: bestBlockHeight } } = await this.coreMethods.getStatus();
+      const createHistoricalSyncStream = (fromBlockHeight, count) => {
+        const { subscribeToBlockHeadersWithChainLocks } = this.coreMethods;
+        return subscribeToBlockHeadersWithChainLocks({
+          fromBlockHeight,
+          count,
+        });
+      };
 
-    if (!this.blockHeadersReader) {
       this.blockHeadersReader = new BlockHeadersReader(
         {
           coreMethods: this.coreMethods,
           maxParallelStreams: this.options.maxParallelStreams,
           targetBatchSize: this.options.targetBatchSize,
           maxRetries: this.options.maxRetries,
+          createContinuousSyncStream,
+          createHistoricalSyncStream,
         },
       );
     }
 
-    this.blockHeadersReader.on(BlockHeadersReader.EVENTS.ERROR, (e) => {
-      this.emit(EVENTS.ERROR, e);
-    });
+    this.blockHeadersReader.on(BlockHeadersReader.EVENTS.BLOCK_HEADERS, this.handleHeaders);
+    this.blockHeadersReader.on(BlockHeadersReader.EVENTS.ERROR, this.handleError);
+  }
 
-    this.blockHeadersReader.on(BlockHeadersReader.EVENTS.HISTORICAL_DATA_OBTAINED, () => {
-      this.blockHeadersReader.subscribeToNew(bestBlockHeight)
-        .catch((e) => {
-          this.emit(EVENTS.ERROR, e);
-        });
-    });
-
-    this.blockHeadersReader.on(BlockHeadersReader.EVENTS.BLOCK_HEADERS, (headers, reject) => {
-      try {
-        this.spvChain.addHeaders(headers.map((header) => Buffer.from(header)));
-      } catch (e) {
-        if (e.message === 'Some headers are invalid') {
-          reject(e);
-        } else {
-          this.emit(EVENTS.ERROR, e);
-        }
-      }
-    });
+  destroyReader() {
+    this.removeReaderListeners();
+    this.blockHeadersReader = null;
+  }
+
+  /**
+   * Checks whether spv chain has header at specified height and flushes chains if not
+   *
+   * @private
+   * @param height
+   */
+  ensureChainRoot(height) {
+    // Flush spv chain in case header at specified height was not found
+    if (!this.spvChain.hashesByHeight[height - 1]) {
+      this.spvChain.reset(height);
+    }
+  }
+
+  /**
+   * Reads historical block headers
+   *
+   * @param fromBlockHeight
+   * @param toBlockHeight
+   * @returns {Promise<void>}
+   */
+  async readHistorical(fromBlockHeight = 1, toBlockHeight) {
```

Comment:
Why the first argument is optional, but the second one is not?

Reply:
Fixed